src/server/req_jobobit.c

```diff
@@ -2269,24 +2288,6 @@ job_obit(struct resc_used_update *pruu, int stream)
 			(JOB_SVFLG_CHKPT | JOB_SVFLG_ChkptMig)) == 0) {
 			job_attr_def[(int)JOB_ATR_resc_used].at_free(
 				&pjob->ji_wattr[(int)JOB_ATR_resc_used]);
-
-		} else if ((pjob->ji_qs.ji_svrflags & JOB_SVFLG_CHKPT) &&
```

Comment:
Are you sure we don't need this part of the code here? What if checkpoint_abort fails on mom and in post_hold we do not remove CHKPT flag from the job and then the job exits. I think in that case control will come here and will not reply ack to mom.
Since this is a timing issue, have you thought of order of operation where server processes obit before processing holdjob response (when job is ending at the same time it is getting held). These are some tricky scenarios and you may want to try some of those.

Reply:
I had tested with the checkpoint_abort script returning 1 instead of 0. Checkpointing fails, but MoM decides to send an Obit that the server responds and the job is preempted.
line 2269 is when substate is rerun and checkpoint flag is NOT set.
I do not think this is a timing issue, PP-1017 was.
I believe, the testing I have done is enough, if there is a specific scenario you have in your mind, please let me know, will be happy to test it.

Reply:
> I had tested with the checkpoint_abort script returning 1 instead of 0. Checkpointing fails, but MoM decides to send an Obit that the server responds and the job is preempted.

I'm not sure I understand what you're saying.  If preempt_order is "C", and the Checkpoint fails, how does the server respond that the job is preempted...didn't preemption fail?

Reply:
the MoM sends an Obit after the checkpointing fails, the server responds to it, and the job gets preempted. The pre-emption goes well regardless of whether checkpointing was successful because the MoM is sending the Obit.

Reply:
@prakashcv13 after digging through the code with Bhroam, I believe you are only seeing this behavior because you killed all processes but returned a non-zero exit.  Since there nothing else for the job to do it ends, and that's why you see the Obit.  What happens if your checkpoint script only returns non-zero exit (need a long enough job that it also doesn't end by then)?  According to what we saw in the code, Mom should be sending a req_reject and not an Obit.

Reply:
Just checking for the job in the Q state doesn't tell if the job was checkpointed.  You need to see if the exec_vnode attribute is also set.  In any case, this seems wrong.  If the checkpoint fails, then the qhold should be rejected.  Could you test this by hand?  Set up checkpoint where the checkpoint_abort script just returns 1.  Submit a job and then qhold it.  If it goes into the H state, something strange is going on.

Reply:
@lisa-altair and @bhroam - I tried this scenario and you guys are right that checkpointing will fail and so will preemption. But, isn't that an expected behavior?
What we are trying to fix here is what went wrong after the fix of PP-1017. It was the missing ack to the obit which is added now. Rest of the things are as they were.
I believe what we have tested so far is enough.

Reply:
If I do not kill the process, the pre-emption still happens. If the process is left lying around, it doesn't.

Reply:
I think this is a side effect of our communication model between mom and the server.  I think what is happening is the server says, "checkpoint this job".  The checkpoint abort script fails but kills the process.  There are no processes left for the job, so the mom notices this and sends the normal end of job obit to the server.  The server was expecting an obit from the checkpoint so it assumes this is where the obit is from.  Now we have a case where the checkpoint_abort script has failed, but we checkpoint the job anyway.  In any case, this isn't part of your bug fix.  Can you file it?  Thanks again for testing it.